### PR TITLE
App Setup: setup new setup settings page

### PR DESCRIPTION
**Pull request by vovakulikov**
Related design issue:
- #50729
- Figma design: https://www.figma.com/file/8UUuaIUPVJDzp8U9ImNvA7/Connect-Cody-and-polish-setup-experience?node-id=210-513&t=eX8TwV3BHeNjaDhy-0
---

## Context 
This PR is the first PR about the new setup settings page; it adds the page itself and main shell layout and reuses the existing setup wizard component step. It's in the WIP stage, but it can be merged since changes to this PR are app-specific. 

It doesn't work in this PR 
- Page Layout isn't final. Based on the conversation with @almeidapaulooliveira, the page will be changed/simplified if there is a need for more app settings except local repositories (remote repositories will be gone since it's not the app priority) 
-  Local repositories action buttons don't work (the form will be revamped in the following PR)
- "Connect to sourcegraph.com" isn't complete, it's in discussion at the moment (do we need to have an additional UI there, or button in the user menu will be enough) 

## Test plan
- Check that all page item links are working properly
- Check that the current setup wizard doesn't have any regressions

### CODE REVIEW

**Comment by sourcegraph-buildkite** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.05% (+1.26 kb) | 0.07% (+10.66 kb) 🔺 | 0.08% (+9.39 kb) | 0.39% (+3) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/vk/setup-new-setup-settings-page/compare-report.html;tab=live_object) for a full comparison between the commits e541cea and fd327b7 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/web/src/enterprise/app/settings/AppSettingsArea.tsx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+            <Outlet />
+        </div>
+    )
+}
+
+interface LocalRepositoriesTabProps extends TelemetryProps {}
+
+const LocalRepositoriesTab: FC<LocalRepositoriesTabProps> = props => {
+    const { telemetryService } = props
```
Comment (limitedmage) [inline client/web/src/enterprise/app/settings/AppSettingsArea.tsx:85]: suggestion: Simplify code, if we don't foresee adding more props in the very-near-future, I don't think we should have empty interface with only one parent.

```suggestion
const LocalRepositoriesTab: FC<TelemetryProps> = ({ telemetryService }) => {
```

File: client/web/src/enterprise/app/settings/AppSettingsArea.tsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+            <LocalRepositoriesStep description={false} progressBar={false} telemetryService={telemetryService} />
+        </div>
+    )
+}
+
+interface RemoteRepositoriesTabProps extends TelemetryProps {}
+
+const RemoteRepositoriesTab: FC<RemoteRepositoriesTabProps> = props => {
+    const { telemetryService } = props
```
Comment (limitedmage) [inline client/web/src/enterprise/app/settings/AppSettingsArea.tsx:115]: same as above
```suggestion
const RemoteRepositoriesTab: FC<TelemetryProps> = ({ telemetryService }) => {
```

File: client/web/src/enterprise/app/settings/AppSettingsArea.tsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+enum AppSettingURL {
+    LocalRepositories = 'local-repositories',
+    ConnectToSourcegraphDotCom = 'connect-to-dot-com',
+    RemoteRepositories = 'remote-repositories',
+}
+
+interface AppSettingsAreaProps extends TelemetryProps {}
+
+export const AppSettingsArea: FC<AppSettingsAreaProps> = props => {
+    const { telemetryService } = props
```
Comment (limitedmage) [inline client/web/src/enterprise/app/settings/AppSettingsArea.tsx:22]: same as below
```suggestion
export const AppSettingsArea: FC<TelemetryProps> = ({ telemetryService }) => {
```

File: client/web/src/enterprise/user/routes.tsx [hunk 4]
```diff
@@ -33,6 +35,14 @@ export const enterpriseUserAreaRoutes: readonly UserAreaRoute[] = [
     ...userAreaRoutes,
     ...enterpriseNamespaceAreaRoutes,
 
+    // Sourcegraph app specific route (cody/app settings page) this route won't be
+    // available for any non-app deploy types see enterprise userAreaHeaderNavItems
+    // for more context on user settings page
```
Comment (limitedmage) [inline client/web/src/enterprise/user/routes.tsx:40]: nit: this comment was hard to parse for me, here's a suggested rewording.

```suggestion
    // Sourcegraph app specific route (cody/app settings page)
    // This route won't be available for any non-app deploy types.
    // See userAreaHeaderNavItems in client/web/src/enterprise/user/navitems.ts
    // for more context on user settings page.
```

File: client/web/src/enterprise/user/routes.tsx [hunk 5]
```diff
@@ -33,6 +35,14 @@ export const enterpriseUserAreaRoutes: readonly UserAreaRoute[] = [
...
+    {
+        path: 'app-settings/*',
+        render: props => <AppSettingsArea telemetryService={props.telemetryService} />,
```
Comment (limitedmage) [inline client/web/src/enterprise/user/routes.tsx:43]: Can we add a condition here so this returns 404 on non-app builds?

File: client/web/src/setup-wizard/components/index.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+// Re-export setup steps to reuse them in App/Cody settings
+// page settings tabs.
```
Comment (limitedmage) [inline client/web/src/setup-wizard/components/index.ts:2]:
```suggestion
// Re-export setup steps to reuse them in App/Cody settings pages.
```

File: client/web/src/setup-wizard/components/SyncRepositoriesStep.tsx [hunk 7]
```diff
@@ -7,10 +7,12 @@ import { SiteAdminRepositoriesContainer } from '../../site-admin/SiteAdminReposi
...
 import { CustomNextButton } from './setup-steps'
 
-interface SyncRepositoriesStep extends TelemetryProps {}
+interface SyncRepositoriesStep extends TelemetryProps {
+    baseURL: string
+}
 
 export function SyncRepositoriesStep(props: SyncRepositoriesStep): ReactElement {
-    const { telemetryService, ...attributes } = props
+    const { telemetryService, baseURL, ...attributes } = props
```
Comment (limitedmage) [inline client/web/src/setup-wizard/components/SyncRepositoriesStep.tsx:15]: I don't like this style of declaring components and destructuring props, it feels like a source of potential typing mistakes and creates an unnecessary extra `props` variable that we don't use again.

I recommend putting the destructuring in the declaration, declaring the component with a specific type, and deleting the line below (GitHub won't let me suggest that 🤷‍♀️)

```suggestion
export const SyncRepositoriesStep: FC<SyncRepositoriesStep> = ({ telemetryService, baseURL, ...attributes }) => {
```

File: client/web/src/setup-wizard/components/local-repositories-step/LocalRepositoriesStep.tsx [hunk 8]
```diff
@@ -43,10 +43,13 @@ import { DISCOVER_LOCAL_REPOSITORIES, GET_LOCAL_CODE_HOSTS, GET_LOCAL_DIRECTORY_
...
 
-interface LocalRepositoriesStepProps extends TelemetryProps, HTMLAttributes<HTMLDivElement> {}
+interface LocalRepositoriesStepProps extends TelemetryProps, HTMLAttributes<HTMLDivElement> {
+    description?: boolean
+    progressBar?: boolean
+}
 
 export const LocalRepositoriesStep: FC<LocalRepositoriesStepProps> = props => {
-    const { telemetryService, ...attributes } = props
+    const { telemetryService, description = true, progressBar = true, ...attributes } = props
```
Comment (limitedmage) [inline client/web/src/setup-wizard/components/local-repositories-step/LocalRepositoriesStep.tsx:52]: Same comment here re: moving prop destructuring to the line above

File: client/web/src/setup-wizard/components/remote-repositories-step/RemoteRepositoriesStep.tsx [hunk 9]
```diff
@@ -21,16 +21,20 @@ import { GET_CODE_HOSTS } from './queries'
...
-interface RemoteRepositoriesStepProps extends TelemetryProps, HTMLAttributes<HTMLDivElement> {}
+interface RemoteRepositoriesStepProps extends TelemetryProps, HTMLAttributes<HTMLDivElement> {
+    baseURL: string
+    description?: boolean
+    progressBar?: boolean
+}
 
 export const RemoteRepositoriesStep: FC<RemoteRepositoriesStepProps> = props => {
-    const { className, telemetryService, ...attributes } = props
+    const { className, telemetryService, baseURL, description = true, progressBar = true, ...attributes } = props
```
Comment (limitedmage) [inline client/web/src/setup-wizard/components/remote-repositories-step/RemoteRepositoriesStep.tsx:31]: Same comment here re: moving prop destructuring to the line above

**Review by limitedmage — APPROVED** [review]
Looks good! Left some suggestions on code style.

**Review by limitedmage — APPROVED** [review]
